### ISSUE TITLE
How $env:PSModulePath is used to find modules should be explained

### ISSUE THREAD

**Issue by jazzdelightsme** [MicrosoftDocs/PowerShell-Docs#6867]
I thought that if I added a path like "C:\src" to $env:PSModulePath, then it would make modules such as the following automatically discoverable/loadable:

    C:\src\foo\foo.psd1
    C:\src\foo\1.2.3\foo.psd1

And indeed, that works. But what I did NOT know is that PSModulePath apparently does a complete subtree search of any dir you put into it. For example, this module would ALSO become available:

    C:\src\fee\fie\foe\fum\fum.psm1

I did not expect that.

We should document how PowerShell uses the value of $env:PSModulePath to find modules.

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 9e779ca4-dbd0-480a-ccb8-ee65df2da1c0
* Version Independent ID: 0e9c150c-ad17-86a9-f0e1-68c90cd93ead
* Content: [about_PSModulePath - PowerShell](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_psmodulepath?view=powershell-7)
* Content Source: [reference/7.0/Microsoft.PowerShell.Core/About/about_PSModulePath.md](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/7.0/Microsoft.PowerShell.Core/About/about_PSModulePath.md)
* Product: **powershell**
* Technology: **powershell-conceptual**
* GitHub Login: @SteveL-MSFT
* Microsoft Alias: **slee**

**Comment by rjmholt** [comment]
> But what I did NOT know is that PSModulePath apparently does a complete subtree search of any dir you put into it. For example, this module would ALSO become available

Testing this, it seems this is indeed the case, but it's contrary to my expectation:

```powershell
mkdir '~\Documents\PowerShell\Modules\fee\fie\foe\fum'

@'
function Test-FeeFieFoeFum
{
    "I smell the blood of an Englishman"
}
'@ > '~\Documents\PowerShell\Modules\fee\fie\foe\fum\fum.psm1'

gmo -list # Includes fum.psm1
```

```
    Directory: C:\Users\Robert Holt\Documents\PowerShell\Modules\fee\fie\foe

ModuleType Version    PreRelease Name                                PSEdition ExportedCommands
---------- -------    ---------- ----                                --------- ----------------
Script     0.0                   fum                                 Desk      Test-FeeFieFoeFum
```

That's pretty worrying to me. It implies we're working way too hard, but also that modules that users wouldn't expect to be exposed are. This feels like a PowerShell issue to me.

However, I get the same behaviour with Windows PowerShell 5.1, so clearly, intended or not, it's been in there for a long time.

Maybe @daxian-dbw or @BrucePay knows more?

**Comment by SteveL-MSFT** [comment]
I agree that it's probably not expected to be recursive and doing more work than necessary although it's also unlikely that users are adding top level paths to the env var and happen to have deep directory structures.  Since this is not a breaking change from 5.1, we should just document this behavior.

**Comment by jazzdelightsme** [comment]
>  it's also unlikely that users are adding top level paths to the env var and happen to have deep directory structures

The reason I asked is because I just wrote an install script thingy that will add a path to the env. var. And the user is given the choice to choose a dir themselves if they like. I added a check to make sure that they wouldn't be inadvertently making *other* modules automatically available (besides the one being installed)... but it turns out my check did not go far enough--and has in fact caused other modules to be made available that weren't intended. :/
